frontend/catalyst/compiler.py

```diff
@@ -268,6 +270,60 @@ def get_output_filename(infile):
         return infile.replace(".ll", ".o")
 
 
+class WrapperToCatchExceptions(PassPipeline):
+    """Create a wrapper around the generated function to catch runtime exceptions."""
+
+    def __init__(self, params, name):
+        self.params = params
+        self.name = name
+
+    _include = sysconfig.get_paths()["include"]
+    _executable = "clang++"
+    _default_flags = ["-fPIC", "-x", "c++", f"-I{_include}"]
+
+    @staticmethod
+    def get_output_filename(infile):
+        # Technically we are just using this to get the whole path
+        if not infile.endswith(".mlir"):
+            raise ValueError("Input is not an MLIR file.")
+        return infile.replace(".mlir", ".wrap.o")
+
+    def run(self, infile, outfile=None, flags=None, options=None):
+        template = """
+#define PY_SSIZE_T_CLEAN
+#include <Python.h>
+#include <stdexcept>
+#include <stdarg.h>
+
+extern "C" {
+  extern void _mlir_ciface_{{name}} ({% for variable in range(variables) %}void*{% if not loop.last %}, {% endif %}{% endfor %});
+  void wrapper_mlir_ciface_{{name}} ({% for variable in range(variables) %}void*{% if not loop.last %}, {% endif %}{% endfor %});
+}
+
+void wrapper_mlir_ciface_{{name}} ({% for variable in range(variables) %}void *arg{{variable}}{% if not loop.last %}, {% endif %}{% endfor %}) {
+  try {
+    _mlir_ciface_{{name}} ({% for variable in range(variables) %}arg{{variable}}{% if not loop.last %}, {% endif %}{% endfor %});
+  } catch (const std::exception &ex) {
+    PyErr_SetString(PyExc_RuntimeError, ex.what());
+  } catch (const std::logic_error &le) {
+    PyErr_SetString(PyExc_RuntimeError, le.what());
+  } catch (const std::range_error &re) {
+    PyErr_SetString(PyExc_RuntimeError, re.what());
+  }
+}
+        """
+        data = {"variables": self.params, "name": self.name}
+        j2_template = Template(template)
+        cooked_template = j2_template.render(data)
```

Comment:
How complex of a template use case do we have here and do you think it justifies and additional dependency?

Reply:
I wasn't making any statement one way or another, just asking us to evaluate the use of the external dependency 🙂 

I wouldn't advocate for us implementing any special purpose functions, but if the template merely consists of straightforward variable substitution (which I'm not saying that it does, hence the request for evaluation), then that is built directly into Python.

Reply:
The template is not just variable substitution (although any template could be variable substitution if one pre-computes the strings correctly before substituting). I think embedding the template logic into the template (which is what Jinja2 does) is better than pre-computing the strings and then using python f-strings. 

Here is how one would change the template to use f-strings.

```python
params = [ f"void* arg{x}" for x in range(num_params)]
args = [ f"arg{x}" for x in range(num_params)]
param_string = ", ".join(params)
args_string = ", ".join(args)
f"""
...{params_string}
...{args_string}
"""
```

Advocating from a user stand point, I agree that this could be implemented with f-strings. But I think that this is a relatively minor dependency that I would be happy to have.

If you insist, I will remove the dependency.

Reply:
Ah, I see. I thought we used more dependencies. (I suppose transitive dependencies and build dependencies don't count) happy to remove it to keep it in as count == 2.

Reply:
Just in case you want to see the concrete impact of this change: https://github.com/PennyLaneAI/catalyst/pull/97/commits/392a9a77a8f69544508a938a94fbebe5ae6fa0e0